File: app/modules/bigboard.py
"""Tab 2 — Rankings Comparison: consensus vs every board, disagreement."""
from __future__ import annotations

import pandas as pd
from shiny import module, reactive, render, ui

from logic import dataio
from modules.player_modal import player_modal

_DF = dataio.consensus()
_KEYS = dataio.source_keys()


def _cell(src_rank, consensus_rank) -> str:
    if src_rank is None or (isinstance(src_rank, float) and pd.isna(src_rank)):
        return '<td class="bb-na">·</td>'
    src_rank = int(src_rank)
    diff = consensus_rank - src_rank
    cls = "bb-hi" if diff >= 15 else ("bb-lo" if diff <= -15 else "")
    return f'<td class="{cls}">{src_rank}</td>'


@module.ui
def bigboard_ui():
    return ui.div(
        ui.input_slider("topn", "Players to show (by consensus rank)", 10,
                        max(len(_DF), 20), min(40, max(len(_DF), 20))),
        ui.help_text("Green = this board is notably higher on the player than consensus; "
                     "red = notably lower. '·' = not ranked by that board."),
        ui.output_ui("matrix"),
        ui.h3("Biggest disagreements"),
        ui.help_text("Players the boards most disagree on — standard deviation of rank across boards."),
        ui.output_data_frame("disagree"),
        ui.output_ui("modal_host"),
    )


@module.server
def bigboard_server(input, output, session):
    input_id = session.ns("select_player")

    @render.ui
    def matrix():
        if not len(_DF):
            return ui.p("No data available.", class_="muted")
        d = _DF.nsmallest(input.topn(), "consensus_rank")
        head = "".join(
            f'<th class="bb-src">{dataio.source_label(k)}</th>' for k in _KEYS
        )
        body = []
        for _, r in d.iterrows():
            cr = int(r["consensus_rank"])
            cells = "".join(_cell(r.get(f"src_{k}"), cr) for k in _KEYS)
            safe_name = str(r["player"]).replace("'", "\\'").replace('"', '\\"')
            body.append(
                f'<tr><td class="bb-rank">{cr}</td>'
                f'<td class="bb-name" style="cursor:pointer" '
                f'onclick="Shiny.setInputValue(\'{input_id}\', \'{safe_name}\', {{priority:\'event\'}});">'
                f'{r["player"]}</td>'
                f'<td class="bb-pos">{r["position"]}</td>{cells}'
                f'<td class="bb-sd">{float(r["stdev"]):.1f}</td></tr>'
            )
        return ui.HTML(
            '<div class="bb-wrap"><table class="bb-table"><thead><tr>'
            '<th>Rank</th><th>Player</th><th>Pos</th>' + head +
            '<th>SD</th></tr></thead><tbody>' + "".join(body) + "</tbody></table></div>"
        )

    @render.data_frame
    def disagree():
        if not len(_DF):
            return render.DataGrid(pd.DataFrame())
        d = _DF[_DF["n_sources"] >= 2].nlargest(40, "stdev")
        cols = ["consensus_rank", "player", "position", "school",
                "best_rank", "worst_rank", "spread", "stdev", "n_sources"]
        return render.DataGrid(d[cols], selection_mode="row", height="420px", width="100%")

    @render.ui
    def modal_host():
        # Name click from matrix table
        try:
            name = input.select_player()
            if name:
                ui.modal_show(player_modal(name))
                return ui.div()
        except Exception:
            pass
        # Row selection from disagree grid
        sel = disagree.data_view(selected=True)
        if sel is None or not len(sel):
            return ui.div()
        ui.modal_show(player_modal(sel.iloc[0]["player"]))
        return ui.div()

File: static/js/bigboard.js
const SOURCE_LABELS = {
  mlb_pipeline: 'MLB Pipeline', baseball_america: 'BA', espn_mcdaniel: 'ESPN/McDaniel',
  keith_law: 'Keith Law', just_baseball: 'Just Baseball', perfect_game: 'Perfect Game',
  fangraphs: 'FanGraphs', overslot: 'Overslot', eleven_point7: '11point7',
};
const SRC_PRIORITY = ['mlb_pipeline','baseball_america','overslot'];

function esc(s) {
  return String(s??'').replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;').replace(/"/g,'&quot;');
}
function srcLabel(k) {
  return SOURCE_LABELS[k] || k.replace(/_/g,' ').replace(/\b\w/g,c=>c.toUpperCase());
}

export function initBigboard(G, showModal) {
  const el = document.getElementById('tab-rankings');
  const players = G.consensus;
  const keys = G.source_keys;

  el.innerHTML = `
    <div style="margin-bottom:1rem;">
      <label class="control-label">Players to show (by consensus rank)</label>
      <input type="range" id="bb-topn" min="10" max="${Math.max(players.length,20)}" value="${Math.min(40,players.length)}" style="width:260px;vertical-align:middle;margin-left:.5rem;">
      <span id="bb-topn-val" style="font-family:'JetBrains Mono',monospace;margin-left:.4rem;">${Math.min(40,players.length)}</span>
    </div>
    <p class="muted" style="font-size:.82rem;margin-bottom:.6rem;">Green = this board notably higher than consensus; red = notably lower. · = not ranked.</p>
    <div id="bb-matrix"></div>
    <h3 style="margin-top:1.5rem;">Biggest disagreements</h3>
    <p class="muted" style="font-size:.82rem;margin-bottom:.6rem;">Players the boards most disagree on — standard deviation of rank across boards.</p>
    <div id="bb-disagree"></div>
  `;

  function renderMatrix(topn) {
    const d = players.slice(0, topn);
    const head = `<th>Rank</th><th>Player</th><th class="bb-pos">Pos</th>` +
      keys.map(k => `<th class="bb-src">${esc(srcLabel(k))}</th>`).join('') +
      `<th class="bb-sd">SD</th>`;
    const body = d.map(p => {
      const cr = p.consensus_rank;
      const cells = keys.map(k => {
        const v = p.sources[k];
        if (v == null) return `<td class="bb-na">·</td>`;
        const diff = cr - v;
        const cls = diff >= 15 ? 'bb-hi' : diff <= -15 ? 'bb-lo' : '';
        return `<td class="${cls}">${v}</td>`;
      }).join('');
      return `<tr>
        <td class="bb-rank">${cr}</td>
        <td class="bb-name" style="cursor:pointer" data-pid="${esc(p.player_id)}">${esc(p.player)}</td>
        <td class="bb-pos">${esc(p.position)}</td>
        ${cells}
        <td class="bb-sd">${p.stdev.toFixed(1)}</td>
      </tr>`;
    }).join('');
    document.getElementById('bb-matrix').innerHTML =
      `<div class="bb-wrap"><table class="bb-table"><thead><tr>${head}</tr></thead><tbody>${body}</tbody></table></div>`;
    document.getElementById('bb-matrix').querySelectorAll('[data-pid]').forEach(el => {
      el.addEventListener('click', () => showModal(el.dataset.pid));
    });
  }

  function renderDisagree() {
    const d = players.filter(p => p.n_sources >= 2).sort((a,b) => b.stdev - a.stdev).slice(0,40);
    const head = `<tr><th>Cons.</th><th>Player</th><th>Pos</th><th>School</th><th>Best</th><th>Worst</th><th>Spread</th><th>SD</th><th>Boards</th></tr>`;
    const body = d.map(p => `<tr style="cursor:pointer" data-pid="${esc(p.player_id)}">
      <td class="bb-rank">${p.consensus_rank}</td>
      <td class="bb-name">${esc(p.player)}</td>
      <td class="bb-pos">${esc(p.position)}</td>
      <td style="text-align:left;font-size:.82rem;">${esc(p.school)}</td>
      <td>${p.best_rank}</td><td>${p.worst_rank}</td><td>${p.worst_rank - p.best_rank}</td>
      <td class="bb-sd">${p.stdev.toFixed(1)}</td><td>${p.n_sources}</td>
    </tr>`).join('');
    document.getElementById('bb-disagree').innerHTML =
      `<div class="bb-wrap"><table class="bb-table"><thead>${head}</thead><tbody>${body}</tbody></table></div>`;
    document.getElementById('bb-disagree').querySelectorAll('[data-pid]').forEach(el => {
      el.addEventListener('click', () => showModal(el.dataset.pid));
    });
  }

  const slider = el.querySelector('#bb-topn');
  const label  = el.querySelector('#bb-topn-val');
  slider.addEventListener('input', () => { label.textContent = slider.value; renderMatrix(+slider.value); });

  renderMatrix(+slider.value);
  renderDisagree();
}
